mm2src/kdf_walletconnect/src/session.rs

```diff
@@ -0,0 +1,378 @@
+    active_topic: Mutex<Option<Topic>>,
+    /// A thread-safe map of sessions indexed by topic.
+    sessions: DashMap<Topic, Session>,
+    pub(crate) storage: SessionStorageDb,
+}
+
+pub struct SessionManager(Arc<SessionManagerImpl>);
+
+impl From<Session> for SessionRpcInfo {
+    fn from(value: Session) -> Self {
+        Self {
+            topic: value.topic,
+            metadata: value.controller.metadata,
+            peer_pubkey: value.controller.public_key,
+            pairing_topic: value.pairing_topic,
+            namespaces: value.namespaces,
+            subscription_id: value.subscription_id,
+            properties: value.session_properties,
+            expiry: value.expiry,
+        }
+    }
+}
+
+#[allow(unused)]
+impl SessionManager {
+    pub(crate) fn new(storage: SessionStorageDb) -> Self {
+        Self(
+            SessionManagerImpl {
+                active_topic: Default::default(),
+                sessions: Default::default(),
+                storage,
+            }
+            .into(),
+        )
+    }
+
+    pub(crate) fn storage(&self) -> &SessionStorageDb { &self.0.storage }
+
+    /// Get active session topic or return error if no session has been activated.
+    pub async fn get_active_topic_or_err(&self) -> MmResult<Topic, WalletConnectError> {
+        self.0
+            .active_topic
+            .lock()
+            .await
+            .clone()
+            .ok_or(MmError::new(WalletConnectError::SessionError(
+                "No active session".to_owned(),
+            )))
+    }
+
+    /// Inserts `Session` into the session store, associated with the specified topic.
+    /// If a session with the same topic already exists, it will be overwritten.
+    pub(crate) async fn add_session(&self, session: Session) {
+        // set active session topic.
+        *self.0.active_topic.lock().await = Some(session.topic.clone());
+        // insert session
+        self.0.sessions.insert(session.topic.clone(), session);
+    }
+
+    /// Removes session corresponding to the specified topic from the session store.
+    /// If the session does not exist, this method does nothing.
+    pub(crate) async fn delete_session(&self, topic: &Topic) -> Option<Session> {
+        info!("[{topic}] Deleting session with topic");
+        let mut active_topic = self.0.active_topic.lock().await;
+
+        // Remove the session and get the removed session (if any)
+        let removed_session = self.0.sessions.remove(topic).map(|(_, session)| session);
+
+        // Update active topic if necessary
+        if active_topic.as_ref() == Some(topic) {
+            // If the deleted session was the active one, find a new active session
+            *active_topic = self.0.sessions.iter().next().map(|session| session.topic.clone());
+
+            if let Some(new_active_topic) = active_topic.as_ref() {
+                info!("[{new_active_topic}] New session with topic activated!");
+            }
+        }
```

Comment:
do we need to pick a new active session? why can't we leave it None?

Reply:
no big reason tbh just thought it could be a good UX